Automattic/wp-calypso
# ETK: Better webpack and asset loading strategy

_New take at #48896. Mostly the same, but rebased with some improvements._

Note: not sure how deployment will be affected by changing asset locations.

### Changes proposed in this Pull Request
Updates the webpack build in the editing toolkit plugin in a few ways:
1. Removes the need for the various package.json entries for each module.
2. Bundles all assets into a root "dist/" directory, rather than having a separate "dist/" directory for each plugin.
3. Creates a function called `use_webpack_assets`, which allows us to access webpack bundled assets in a standardized manner -- and helps remove several hundred LOC.
4. Updates all modules to load with this new format.

#### Webpack changes

The webpack config changes work like this:
1. Parses the files in the `editing-toolkit-plugin` directory to find each subdirectory which contains an `index.(j|t)s` file.
2. Removes any entries which are part of the `excludedSources` list in package.json
3. Adds other sources from `additionalSources`
4. Standardizes everything such that all sources are compiled into the top-level of the "dist" directory.

#### Asset enqueue changes
I'd been wanting to write something like this for a while, because I've noticed that projects setting up in the plugin tend to do things slightly differently with this part of the code. It is about 20 LOC which needs to be repeated for every project for setup, so this improves things a lot:

1. Less code to maintain
2. You only have to remember one function to use when setting up new code.
3. All code is registered similarly, so things like translations work "out of the box".

#### Testing instructions
* Test every feature of the plugin again
* Test on your sandbox with `yarn dev --sync` locally to test the webpack changes.
* Test the production build with `install-plugin.sh etk etk/webpack-improvements`
* Verify PHPunit tests with `yarn test:php` from etk directory (make sure wp-env is started)
